code/parachain/frame/composable-maths/src/dex/constant_product.rs

```diff
@@ -123,6 +123,77 @@ where
 	Ok(ai)
 }
 
+/// Compute the amount of the input token given the amount of the output token.
+///
+/// If `Ok`, returns a tuple containing `(a_sent, fee)`.
+/// To get `a_sent` without accounting for the fee, set `f = 0`.
+/// Amount in, round up results.
+///
+/// **NOTE:** Weights must already be normalized.
+///
+/// # Parameters
+/// * `w_i` - Weight of the input token
+/// * `w_o` - Weight of the output token
+/// * `b_i` - Balance of the input token
+/// * `b_o` - Balance of the output token
+/// * `a_out` - Amount of the output token desired by the user
+/// * `f` - Total swap fee
+///
+/// From https://github.com/ComposableFi/composable/blob/main/rfcs/0008-pablo-lbp-cpp-restructure.md#41-fee-math-updates,
+/// equation (3)
+pub fn compute_in_given_out_new<T: PerThing>(
+	w_i: T,
+	w_o: T,
+	b_i: u128,
+	b_o: u128,
+	a_out: u128,
+	f: T,
+) -> Result<(u128, u128), DispatchError> {
+	ensure!(
+		a_out <= b_o,
+		DispatchError::from(
+			"`a_out` must not be graeter than `b_o` (can't take out more than what's available)"
+		)
+	);
+	println!("past check!!!!!!!!!");
+	let w_i = Decimal::from(w_i.deconstruct().into());
+	let w_o = Decimal::from(w_o.deconstruct().into());
+	let b_i = Decimal::from(b_i);
+	let b_o = Decimal::from(b_o);
+	let a_out = Decimal::from(a_out);
+
+	let weight_ratio = w_o.safe_div(&w_i)?;
+	// NOTE(connor): Use if to prevent pointless conversions if `f` is zero
+	let left_from_fee = if f.is_zero() {
+		Decimal::ONE
+	} else {
+		Decimal::from(f.left_from_one().deconstruct().into())
+			.safe_div(&Decimal::from(T::one().deconstruct().into()))?
+	};
+	let b_i_over_fee = b_i.safe_div(&left_from_fee)?;
+	let fee = Decimal::ONE.safe_sub(&left_from_fee)?;
+
+	let value = b_o.safe_sub(&a_out)?;
+	let value = b_o.safe_div(&value)?;
+	let value = value.checked_powd(weight_ratio).ok_or(ArithmeticError::Overflow)?;
+	let value = value.safe_sub(&Decimal::ONE)?;
+
+	let a_sent = round_up(b_i_over_fee.safe_mul(&value)?);
+	let fee = round_up(a_sent.safe_mul(&fee)?).to_u128().ok_or(ArithmeticError::Overflow)?;
+
+	Ok((a_sent.to_u128().ok_or(ArithmeticError::Overflow)?, fee))
+}
+
+/// Rounds a decimal value up to the nearest whole number
+fn round_up(decimal: Decimal) -> Decimal {
+	round(decimal, RoundingStrategy::AwayFromZero)
+}
+
+/// Rounds a decimal value to a whole number based on the provided `RoundingStrategy`
+fn round(decimal: Decimal, rounding_strategy: RoundingStrategy) -> Decimal {
+	decimal.round_dp_with_strategy(0, rounding_strategy)
+}
```

Comment:
NIT: I think using the `round_dp_with_strategy` method directly would be more clear than these helper functions

Reply:
I think they are helpful as they abstract all the non standard terminology in this weird library

Reply:
I had it originally, but yeah, lots of library specifics here that I don't think are useful for reading/evaluating the function.

Reply:
I was under the impression that I couldn't extend it because it's not from this lib. I'll double-check, it might be considered a primitive

Reply:
Just a trait with the required methods that you implement for `Decimal` only:

```rust
trait DecimalExt: Sized {
	fn round_up(self) -> Decimal;

	/// Rounds a decimal value to a whole number based on the provided `RoundingStrategy`
	fn round(self, rounding_strategy: RoundingStrategy) -> Decimal;
}

impl DecimalExt for Decimal {
	fn round_up(self) -> Decimal {
		self.round(RoundingStrategy::AwayFromZero)
	}

	/// Rounds a decimal value to a whole number based on the provided `RoundingStrategy`
	fn round(self, rounding_strategy: RoundingStrategy) -> Decimal {
		self.round_dp_with_strategy(0, rounding_strategy)
	}
}
```